torch/overrides.py

```diff
@@ -1766,16 +1766,23 @@ def is_tensor_like(inp):
     return type(inp) is torch.Tensor or hasattr(type(inp), "__torch_function__")
 
 
+_cur_torch_function_mode = []
+
 def _wrap_torch_function(f):
     @functools.wraps(f)
     def wrapped(self, *args, **kwargs):
         if isinstance(f, classmethod):
             raise RuntimeError("TorchFunctionMode's torch_function function " +
                                "should be a normal method not a class method")
-        inner = getattr(self, "inner", None)
 
-        with enable_torch_function_mode(inner):
-            return f(self, *args, **kwargs)
+        inner = _cur_torch_function_mode.pop() if len(_cur_torch_function_mode) > 0 else None
+
+        try:
+            with enable_torch_function_mode(inner):
+                return f(self, *args, **kwargs)
+        finally:
+            assert _get_torch_function_mode() is None
+            _cur_torch_function_mode.append(inner)
```

Comment:
This is also a bit unexpected. This is the "doing it all in Python" strategy. It seems a little overcomplicated to be manually reenabling the inner torch function mode in the wrapper here. Remember what the C++ invariant is: there is one torch function mode, and when we dispatch to it, C++ will disable the mode. Wrap torch function is responsible for reenabling the inner mode.

Now, you are maintaining the mode stack yourself. So all you really need is one torch function mode, that is "do the next mode on the global stack". When you get to the wrap torch function, pop off the stack (that's the mode you're going to run), restore the one single torch function mode if the stack is non-empty, and then *directly call the mode you got off the stack.*

Reply:
Finally getting back to this as other PRs are getting fixed and since the pin got updated

> So all you really need is one torch function mode, that is "do the next mode on the global stack". When you get to the wrap torch function, pop off the stack (that's the mode you're going to run), restore the one single torch function mode if the stack is non-empty, and then directly call the mode you got off the stack

I don't really get this--are you saying that we should introduce a new torch function mode that is always active at the C++ level after the first mode is pushed and is now in charge of the mode stack?  (i.e. instead of Mode1 -> Mode2 it would go ModeStackMode -> Mode1 -> ModeStackMode -> Mode2 with Mode1 and Mode2 never active as the C++ mode)? If so, I'm unclear how we get to Mode1's torch function. Are we calling `Mode1.__torch_function__(...)`?

It feels like this makes it more complicated since the C++ mode is now just a passer mode and doesn't reflect anything that the user actually passed

Also this would mean that we now break enable_torch_function_mode, right? Since that would break the fact that the only mode on the stack ever is ModeStackMode